File: src/main/java/com/kinvey/sample/statusshare/fragments/UpdateEditFragment.java
/** 
 * Copyright (c) 2014 Kinvey Inc.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 * 
 */
package com.kinvey.sample.statusshare.fragments;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.*;
import com.kinvey.android.Client;
import com.kinvey.java.LinkedResources.LinkedFile;
import com.kinvey.java.core.KinveyClientCallback;
import com.kinvey.java.model.KinveyMetaData;
import com.kinvey.sample.statusshare.MainActivity;
import com.kinvey.sample.statusshare.R;
import com.kinvey.sample.statusshare.model.UpdateEntity;

import java.io.*;

/**
 * @author edwardf
 * @since 2.0
 */
public class UpdateEditFragment extends KinveyFragment implements View.OnClickListener {


    private ImageView attachmentImage;
    private AlertDialog mDialog;

    private EditText updateText;
    private Bitmap image;
    private TextView title;
    private TextView attachmentTitle;

    public UpdateEditFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);


        final String[] items = new String[]{"From Camera", "From SD Card"};
        ArrayAdapter<String> adapter = new ArrayAdapter<String>(getActivity(), android.R.layout.select_dialog_item, items);
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

        builder.setTitle("Select Image");
        builder.setAdapter(adapter, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int item) {
                if (item == 0) {
                    ((MainActivity) getActivity()).startCamera();


                    dialog.cancel();
                } else {

                    ((MainActivity) getActivity()).startFilePicker();
                    dialog.cancel();


                }
            }
        });

        mDialog = builder.create();

    }

    @Override
    public void onResume() {
        super.onResume();
        this.image = ((MainActivity) getActivity()).bitmap;


        if (this.image != null) {
            Log.i(Client.TAG, "setting imageview");
            attachmentImage.setBackgroundDrawable(null);
            attachmentImage.setImageBitmap(this.image);
        } else {

            Log.i(Client.TAG, "not setting imageview");
        }
    }

    @Override
    public int getViewID() {
        return R.layout.fragment_write_update;
    }

    @Override
    public void bindViews(View v) {

        attachmentImage = (ImageView) v.findViewById(R.id.preview);
        updateText = (EditText) v.findViewById(R.id.update);
        title = (TextView) v.findViewById(R.id.share_title);
        attachmentTitle = (TextView) v.findViewById(R.id.share_attach_title);
        title.setTypeface(getRoboto());
        attachmentTitle.setTypeface(getRoboto());
        updateText.setTypeface(getRoboto());

        attachmentImage.setOnClickListener(this);
    }


    public void doUpdate() {
        final ProgressDialog progressDialog = ProgressDialog.show(getActivity(), "",
                "Posting. Please wait...", true);

        byte[] byteArray = null;
        if (image != null) {
            ByteArrayOutputStream stream = new ByteArrayOutputStream();
            image.compress(Bitmap.CompressFormat.PNG, 100, stream);
            byteArray = stream.toByteArray();
            try {
                stream.close();
            } catch (IOException e) {

            }
        }


        saveUpdateAttachment(progressDialog, byteArray, getClient().user().getId() + "_attachment_" + System.currentTimeMillis() + ".png");
    }


    public void saveUpdateAttachment(final ProgressDialog progressDialog, byte[] bytes, String filename) {
        UpdateEntity updateEntity = new UpdateEntity(getClient().user().getId());
        updateEntity.setText(updateText.getText().toString());
        updateEntity.getAcl().setGloballyReadable(true);

        android.util.Log.d(Client.TAG, "updateEntity.getMeta().isGloballyReadable() = " + updateEntity.getAcl().isGloballyReadable());

        if (bytes != null && filename != null) {
            Log.i(Client.TAG, "there is an attachment!");
            LinkedFile lf = new LinkedFile(filename);
            lf.addExtra("_public", true);
            KinveyMetaData.AccessControlList acl = new KinveyMetaData.AccessControlList();
            acl.setGloballyReadable(true);
            lf.addExtra("_acl", acl);
            updateEntity.putFile("attachment", lf);
        }
        final ByteArrayInputStream bais = ((bytes == null) ? null : new ByteArrayInputStream(bytes));
        if (bais != null){
            updateEntity.getFile("attachment").setInput(bais);
        }

        getClient().linkedData(MainActivity.COL_UPDATES, UpdateEntity.class).save(updateEntity, new KinveyClientCallback<UpdateEntity>() {

            @Override
            public void onSuccess(UpdateEntity result) {
                if (getActivity() == null){
                    return;
                }
                android.util.Log.d(Client.TAG, "postUpdate: SUCCESS _id = " + result.getId() + ", gr = " + result.getAcl().isGloballyReadable());
                progressDialog.dismiss();

                try {
                    bais.close();
                } catch (Exception e) {
                }

                InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(getActivity().INPUT_METHOD_SERVICE);
                imm.hideSoftInputFromWindow(updateText.getWindowToken(), 0);

                ((MainActivity) getActivity()).bitmap = null;

                if (getActivity() != null) {
                    ((MainActivity)getActivity()).setShareList(null);

                    ((MainActivity)getActivity()).replaceFragment(new ShareListFragment(), false);

//                    ((StatusShare)getSherlockActivity()).removeFragment(UpdateEditFragment.this);
//                    ((StatusShare)((StatusShare) getSherlockActivity()).removeFragment(getSherlockActivity().getSupportFragmentManager().);)
                }
            }

            @Override
            public void onFailure(Throwable e) {
                if (getActivity() == null){
                    return;
                }
                Log.d(Client.TAG, "failed to upload linked app data");
                e.printStackTrace();
                progressDialog.dismiss();
            }

        }, null
        );
//        } else {
//            Log.i(Client.TAG, "there is no attachment");
//        }
    }


    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        menu.clear();
        inflater.inflate(R.menu.menu_edit_share, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_send_post:
                doUpdate();
                break;
        }

        return super.onOptionsItemSelected(item);
    }


    @Override
    public void onClick(View v) {
        if (v == attachmentImage) {
            mDialog.show();

        }
    }
}

File: app/src/main/java/com/kinvey/sample/statusshare/ui/fragments/UpdateEditFragment.kt
/**
 * Copyright (c) 2019 Kinvey Inc.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */
package com.kinvey.sample.statusshare.ui.fragments

import android.R.layout
import android.app.AlertDialog
import android.app.AlertDialog.Builder
import android.app.ProgressDialog
import android.graphics.Bitmap
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.View.OnClickListener
import android.widget.ArrayAdapter
import com.kinvey.android.store.DataStore
import com.kinvey.java.core.KinveyClientCallback
import com.kinvey.java.linkedResources.LinkedFile
import com.kinvey.java.model.KinveyMetaData.AccessControlList
import com.kinvey.java.store.StoreType
import com.kinvey.sample.statusshare.R
import com.kinvey.sample.statusshare.model.UpdateEntity
import com.kinvey.sample.statusshare.utils.BitmapTools
import com.kinvey.sample.statusshare.utils.Constants
import com.kinvey.sample.statusshare.utils.UiUtils
import kotlinx.android.synthetic.main.fragment_write_update.*
import timber.log.Timber
import java.io.ByteArrayInputStream

/**
 * @author edwardf
 * @since 2.0
 */
class UpdateEditFragment : KinveyFragment(), OnClickListener {

    private var mDialog: AlertDialog? = null
    private var image: Bitmap? = null
    private var dataStore: DataStore<UpdateEntity>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        dataStore = DataStore.collection(Constants.COL_UPDATES, UpdateEntity::class.java, StoreType.AUTO, client)
        setHasOptionsMenu(true)
        val items = arrayOf("From Camera", "From SD Card")
        val adapter = ArrayAdapter(activity!!, layout.select_dialog_item, items)
        val builder = Builder(activity)
        builder.setTitle("Select Image")
        builder.setAdapter(adapter) { dialog, item ->
            if (item == 0) {
                mainActivity?.startCamera()
                dialog.cancel()
            } else {
                mainActivity?.startFilePicker()
                dialog.cancel()
            }
        }
        mDialog = builder.create()
    }

    override fun onResume() {
        super.onResume()
        image = mainActivity?.bitmap
        if (image != null) {
            Timber.i("setting imageview")
            previewImage?.setBackgroundResource(0)
            previewImage?.setImageBitmap(image)
        } else {
            Timber.i("not setting imageview")
        }
    }

    override val viewID = R.layout.fragment_write_update

    override fun bindViews(v: View) {
        previewImage?.setOnClickListener(this)
    }

    fun doUpdate() {
        val progressDialog = ProgressDialog.show(activity, "", "Posting. Please wait...", true)
        val byteArray = BitmapTools.compressImage(image)
        val fileName = "${client?.activeUser?.id}_attachment_${System.currentTimeMillis()}.png"
        saveUpdateAttachment(progressDialog, byteArray, fileName)
    }

    private fun saveUpdateAttachment(progressDialog: ProgressDialog, bytes: ByteArray?, filename: String?) {
        val updateEntity = UpdateEntity(client?.activeUser?.id)
        updateEntity.text = updateEdit.text.toString()
        updateEntity.acl?.setGloballyReadable(true)
        Timber.d("updateEntity.getMeta().isGloballyReadable() = ${updateEntity.acl?.isGloballyReadable}")
        if (bytes != null && filename != null) {
            Timber.i("there is an attachment!")
            val lf = LinkedFile(filename)
            lf.addExtra("_public", true)
            val acl = AccessControlList()
            acl.setGloballyReadable(true)
            lf.addExtra("_acl", acl)
            updateEntity.putFile("attachment", lf)
        }
        val bais = if (bytes == null) null else ByteArrayInputStream(bytes)
        if (bais != null) {
            updateEntity.getFile("attachment")?.input = bais
        }
        dataStore?.save(updateEntity, object : KinveyClientCallback<UpdateEntity> {
            override fun onSuccess(result: UpdateEntity) {
                if (activity == null) { return }
                Timber.d("postUpdate: SUCCESS _id = ${result.id} gr = ${result.acl?.isGloballyReadable}")
                progressDialog.dismiss()
                try {
                    bais?.close()
                } catch (e: Exception) {
                    Timber.e(e)
                }
                UiUtils.hideKeyboard(activity)
                mainActivity?.bitmap = null
                if (activity != null) {
                    mainActivity?.shareList = null
                    replaceFragment(ShareListFragment(), false)
                }
            }
            override fun onFailure(error: Throwable) {
                if (activity == null) { return }
                Timber.d("failed to upload linked app data")
                progressDialog.dismiss()
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
        inflater.inflate(R.menu.menu_edit_share, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_send_post -> doUpdate()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onClick(v: View) {
        if (v === previewImage) {
            mDialog?.show()
        }
    }
}